File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/SecureAccountViewModel.swift
import Foundation
import Observation

/// Backs the **Secure your account** step of create-onboarding.
///
/// Inserted right after an account/identity is opened (OpenAccount) and
/// before the user lands in the main app. It nudges the user to back up
/// the account they just created, with the cloud option pre-selected,
/// while letting them skip behind a clear warning.
///
/// This VM owns ONLY the small selection + skip-confirmation state and
/// the iCloud-availability detection. The two backup mechanisms it routes
/// to are the EXISTING ones:
///   - cloud  → the WebAuthn-PRF cloud recovery (RecoveryViewModel),
///   - file   → the `.flagshipkey` export (KeyfileExportViewModel /
///     KeyfileExportScreen).
/// Nothing here is rebuilt; the screen wires those in.
@Observable
@MainActor
public final class SecureAccountViewModel {
    /// The three mutually-exclusive choices on the step. `none` is the
    /// pre-selected value only when iCloud is unavailable (so the user
    /// still has file + skip).
    public enum Option: Equatable, Sendable {
        case cloud
        case file
    }

    /// The current selection. Starts pre-selected with `.cloud` when
    /// iCloud is available, otherwise `nil` (nothing pre-selected).
    public var selected: Option?

    /// Whether iCloud is reachable for this device/account. When false
    /// the cloud row is shown disabled with a hint and never pre-selected.
    public let iCloudAvailable: Bool

    /// Drives the "Skip for now" confirmation sheet.
    public var showSkipConfirm: Bool = false

    /// Detects iCloud availability via the iCloud ubiquity token — the
    /// standard, entitlement-safe signal for "is the user signed into
    /// iCloud". Returns false (never crashes / blocks) when iCloud is off.
    public nonisolated static func detectICloudAvailable() -> Bool {
        FileManager.default.ubiquityIdentityToken != nil
    }

    public init(iCloudAvailable: Bool = SecureAccountViewModel.detectICloudAvailable()) {
        self.iCloudAvailable = iCloudAvailable
        // Pre-select the cloud option ONLY when iCloud is available.
        // When it's off we pre-select nothing; the step still works via
        // the file option or skip.
        self.selected = iCloudAvailable ? .cloud : nil
    }

    /// The cloud row is selectable only when iCloud is available.
    public var canSelectCloud: Bool { iCloudAvailable }

    /// Continue acts on the current selection. Enabled whenever SOMETHING
    /// is selected (cloud only when available — `selected` can't be
    /// `.cloud` if it isn't, since selecting it is gated).
    public var canContinue: Bool { selected != nil }

    /// Select the cloud option (no-op when iCloud is unavailable).
    public func selectCloud() {
        guard canSelectCloud else { return }
        selected = .cloud
    }

    /// Select the file-backup option.
    public func selectFile() {
        selected = .file
    }

    /// Begin the skip flow — surfaces the confirmation sheet.
    public func requestSkip() {
        showSkipConfirm = true
    }

    /// Dismiss the skip confirmation without skipping ("Back").
    public func cancelSkip() {
        showSkipConfirm = false
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/SecureAccountViewModel.kt
// "Secure your account" — the SKIPPABLE backup nudge shown once, right
// after a brand-new account is opened (create path only; the recovery /
// "I already have an account" path never sees this). It pre-selects the
// cloud (passkey) option when the device can do it, falls back to a
// file-only / skip state when it can't, and never blocks the user from
// reaching the app.
//
// This VM holds ONLY selection + skip-confirm state. The actual cloud
// ceremony reuses PasskeyRecoveryManager + Recovery + BlockStoreUmkStore
// (exactly as RecoveryScreen does) and the file option reuses the
// existing KeyfileExportScreen — neither is rebuilt here.
//
// Copy is approved + verbatim; mirror it on the other surfaces.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** The two backup methods offered on the step. */
enum class SecureAccountOption { Cloud, File }

class SecureAccountViewModel(
    /** Whether this device can register a PRF passkey for cloud backup.
     *  Injected so tests can pin both branches; the screen computes the
     *  real value from the platform (see [PasskeyAvailability]). */
    val passkeyAvailable: Boolean,
) : ViewModel() {

    // Cloud is PRE-SELECTED by default when it's available; when it
    // isn't, nothing is pre-selected (the step still works via file or
    // skip). The cloud option is rendered disabled in that case.
    private val _selected = MutableStateFlow(
        if (passkeyAvailable) SecureAccountOption.Cloud else null,
    )
    val selected: StateFlow<SecureAccountOption?> = _selected.asStateFlow()

    fun select(option: SecureAccountOption) {
        // The cloud option can't be selected when passkeys are
        // unavailable on this device.
        if (option == SecureAccountOption.Cloud && !passkeyAvailable) return
        _selected.value = option
    }

    // Drives the "Skip for now" confirmation dialog.
    private val _showSkipConfirm = MutableStateFlow(false)
    val showSkipConfirm: StateFlow<Boolean> = _showSkipConfirm.asStateFlow()

    fun requestSkip() { _showSkipConfirm.value = true }
    fun cancelSkip() { _showSkipConfirm.value = false }

    /** Continue is meaningful only with a selection. */
    val canContinue: Boolean
        get() = _selected.value != null

    companion object {
        /** Copy — kept here so tests can assert it and the other
         *  surfaces can mirror it verbatim. */
        const val TITLE = "Secure your account"
        const val BODY =
            "Back up your account now so you can get back in if you lose " +
                "this device. No one — not even us — can recover it for you."

        const val CLOUD_LABEL = "Back up with your passkey"
        const val CLOUD_SUBLABEL =
            "Recover with your Google passkey or password manager."
        const val CLOUD_UNAVAILABLE_HINT =
            "Passkeys aren't available on this device — use a backup file."

        const val FILE_LABEL = "Save a backup file"
        const val FILE_SUBLABEL = "An encrypted .flagshipkey you keep yourself."

        const val CONTINUE = "Continue"
        const val SKIP = "Skip for now"

        const val SKIP_WARNING =
            "Without a backup, losing this device means losing your " +
                "account for good. You can set this up anytime in Settings."
        const val SKIP_CONFIRM = "Skip anyway"
        const val SKIP_BACK = "Back"
    }
}
